library/objective-c/EnvoyEngineImpl.m

```diff
@@ -66,40 +62,51 @@ static envoy_headers toNativeHeaders(EnvoyHeaders *headers) {
   return (envoy_headers){length, header_array};
 }
 
+static const void* ios_http_filter_init(const void *context) {
+  EnvoyHTTPFilterFactory *filterFactory = (__bridge EnvoyHTTPFilterFactory *)context;
+  EnvoyHTTPFilter *filter = filterFactory.create();
+  return CFBridgingRetain(filter);
+}
+
 static envoy_filter_headers_status
-ios_http_filter_on_request_headers(envoy_headers headers, bool end_stream, void *context) {
+ios_http_filter_on_request_headers(envoy_headers headers, bool end_stream, const void *context) {
   // TODO(goaway): optimize unmodified case
-  ios_http_filter_context *c = (ios_http_filter_context *)context;
-  if (c->filter.onRequestHeaders == nil) {
+  EnvoyHTTPFilter *filter = (__bridge EnvoyHTTPFilter *)context;
+  if (filter.onRequestHeaders == nil) {
     return (envoy_filter_headers_status){/*status*/ kEnvoyFilterHeadersStatusContinue,
                                          /*headers*/ headers};
   }
 
   EnvoyHeaders *platformHeaders = to_ios_headers(headers);
   release_envoy_headers(headers);
   // TODO(goaway): consider better solution for compound return
-  NSArray *result = c->filter.onRequestHeaders(platformHeaders, end_stream);
+  NSArray *result = filter.onRequestHeaders(platformHeaders, end_stream);
   return (envoy_filter_headers_status){/*status*/ [result[0] intValue],
                                        /*headers*/ toNativeHeaders(result[1])};
 }
 
 static envoy_filter_headers_status
-ios_http_filter_on_response_headers(envoy_headers headers, bool end_stream, void *context) {
+ios_http_filter_on_response_headers(envoy_headers headers, bool end_stream, const void *context) {
```

Comment:
I think we talked about it in one of the earlier PRs, but could we delete the duplication and dispatch to common code between request and reply?

Reply:
I did delete the (more significant) duplication on that PR (in filter.cc). Here, I'd suggest it's pretty limited. We check if the function's present, map the headers, call the function. I almost feel the extra indirection from factoring here would serve only to complicate the code.

Reply:
But I mean, only just almost? If you feel strongly, I can still do it.